modules/javafx.graphics/src/main/java/com/sun/prism/j2d/PrismPrintPipeline.java

```diff
@@ -116,17 +119,91 @@ public synchronized ObservableSet<Printer> getAllPrinters() {
             for (int i=0; i<allServices.length;i++) {
                 if (defService != null && defService.equals(allServices[i])) {
                     printers.add(defPrinter);
+                    pMap.put(defService, defPrinter);
                 } else {
-                    PrinterImpl impl = new J2DPrinter(allServices[i]);
-                    Printer printer = PrintHelper.createPrinter(impl);
-                    impl.setPrinter(printer);
-                    printers.add(printer);
+                    addNew(allServices[i], printers);
                 }
             }
-            printerSet =
-                FXCollections.unmodifiableObservableSet
-                   (FXCollections.observableSet(printers));
+            printerSet = FXCollections.observableSet(printers);
+            returnedPrinterSet =
+                FXCollections.unmodifiableObservableSet(printerSet);
+        } else {
+            PrintService[] newServices =
+                    PrintServiceLookup.lookupPrintServices(null, null);
+            if ((newServices.length != printerSet.size()) ||
+                    (lastTime + 120000) > System.currentTimeMillis()) {
+                updatePrinters(newServices);
+            }
+        }
+        lastTime = System.currentTimeMillis();
```

Comment:
Is `lastTime` intended to be the last time you updated the printers or the last time the `getAllPrinters` method was called? If the former, then you would want to set `lastTime` right after `updatePrinters`. If the latter, then it's fine where it is.

Reply:
The last time we updated.
If the number of printers has changed I will always rebuild the printer list (set) but if they are the same
what are the odds a printer was removed and another one added ? 
I figured if 2 minutes has passed since we last checked it is worth it, but I am not re-setting this every
time we called getAllPrinters() else if you kept calling it every 30 seconds .. then we'd never get past that check.

Reply:
In that case shouldn't the assignment of `lastTime` be moved up two lines, right below the call to `updatePrinters`? Or am I  missing something?

Reply:
I guess I knew what I wanted but didn't have the code to match :-)

Reply:
I think I know how it ended up outside the conditions .. I also need to set it up on the first time through
else the 2nd call will always trigger a refresh since otherwise lastTime will be zero .. 
fixing that.